internal/authentication/oidc/clientcredentials.go

```diff
@@ -0,0 +1,213 @@
+			case <-c.clock.After(renewDuration):
+			}
+
+			c.log.Debug("Renewing client credentials token")
+
+			token, err := c.fetchTokenFn(context.WithValue(ctx, oauth2.HTTPClient, c.httpClient))
+			if err != nil {
+				c.log.Errorf("Error fetching renewed oidc client_credentials token, retrying in 30 seconds: %s", err)
+				renewDuration = time.Second * 30
+				continue
+			}
+
+			c.lock.Lock()
+			c.currentToken = token
+			c.lock.Unlock()
+			renewDuration = c.tokenRenewDuration()
+		}
+	}()
+}
+
+func toConfig(opts ClientCredentialsOptions) (*ccreds.Config, *http.Client, error) {
+	scopes := opts.Scopes
+	if len(scopes) == 0 {
+		// If no scopes are provided, then the default is to use the 'openid' scope
+		// since that is always required for OIDC so implicitly add it.
+		scopes = []string{oidcScopeOpenID}
+	}
+
+	var oidcScopeFound bool
+	for _, scope := range scopes {
+		if scope == oidcScopeOpenID {
+			oidcScopeFound = true
+			break
+		}
+	}
+	if !oidcScopeFound {
+		return nil, nil, fmt.Errorf("oidc client_credentials token source requires the %q scope", oidcScopeOpenID)
+	}
+
+	tokenURL, err := url.Parse(opts.TokenURL)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing token URL: %w", err)
+	}
+	if tokenURL.Scheme != "https" {
+		return nil, nil, fmt.Errorf("OIDC token provider URL requires 'https' scheme: %q", tokenURL)
+	}
+
+	conf := &ccreds.Config{
+		ClientID:     opts.ClientID,
+		ClientSecret: opts.ClientSecret,
+		TokenURL:     opts.TokenURL,
+		Scopes:       scopes,
+	}
+
+	if len(opts.Audiences) == 0 {
+		return nil, nil, errors.New("oidc client_credentials token source requires at least one audience")
+	}
+
+	conf.EndpointParams = url.Values{"audience": opts.Audiences}
+
+	// If caPool is nil, then the Go TLS library will use the system's root CA.
+	var caPool *x509.CertPool
+	if len(opts.CAPEM) > 0 {
+		caPool = x509.NewCertPool()
+		if !caPool.AppendCertsFromPEM(opts.CAPEM) {
+			return nil, nil, errors.New("failed to parse CA PEM")
+		}
+	}
+
+	return conf, &http.Client{
+		Timeout: time.Second * 30,
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				MinVersion: tls.VersionTLS12,
+				RootCAs:    caPool,
+			},
+		},
+	}, nil
+}
+
+func (c *ClientCredentials) Close() {
+	defer c.wg.Wait()
+
+	if c.closed.CompareAndSwap(false, true) {
+		close(c.closeCh)
+	}
+}
+
+func (c *ClientCredentials) Token() (string, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	if c.closed.Load() {
+		return "", errors.New("client_credentials token source is closed")
+	}
+
+	if !c.currentToken.Valid() {
+		return "", errors.New("client_credentials token source is invalid")
+	}
+
+	return c.currentToken.AccessToken, nil
+}
+
+// tokenRenewTime returns the duration when the token should be renewed, which is
+// half of the token's lifetime.
+func (c *ClientCredentials) tokenRenewDuration() time.Duration {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.currentToken.Expiry.Sub(c.clock.Now()) / 2
```

Comment:
No need to halve the time, that's wasteful. Maybe subtract 1 min, but since tokens are used right away there's no need to halve this.

Reply:
How about 2/3rds? I think 1 minute is far too short- node migrations in kube can take longer than a minute for example.

Reply:
Most IdP's I'm familiar with issue tokens with a lifetime of 1h by default. Cutting 30 or 20 mins out of that seems unnecessary.

5 mins?

Why are Kubernetes node migrations a concern here?

Reply:
> It's very common to be running an internal provider in (or in another) cluster. A node migration is an example scenario where the issuer server becomes unavailable for a non-trivial amount of time, however the system should be able to cope with that as best it can.

That is an operational thing that should not be a concern of Dapr's. If an ops person runs their own IdP, they are responsible for ensuring its availability.

Reply:
> That is an operational thing that should not be a concern of Dapr's. If an ops person runs their own IdP, they are responsible for ensuring its availability.

I agree, but it is standard practice as a client to have enough buffer time in the validity of your identity documents to maximize up time if the provider goes down. Using the document validity duration is a great metric to choose that buffer time, rather than using an arbitrary static value.